caffeine@patapon.info/extension.js

```diff
@@ -127,12 +106,302 @@ const AppsTrigger = {
+        let activeWorkspace = global.workspace_manager.get_active_workspace();
+
+        for (let appId of possibleTriggerApps) {
+            let app = this._appSystem.lookup_app(appId);
+            if (app !== null) {
+                if (app.is_on_workspace(activeWorkspace)) {
+                    return appId;
+                }
+            }
+        }
+
+        return null;
+    }
+
+    _getInhibitReasons() {
+        let reasons = [];
+        if (this.isFullscreen() && this._settings.get_boolean(FULLSCREEN_KEY)) {
+            reasons.push('fullscreen');
+        }
+
+        if (this._userEnabled) {
+            reasons.push('user');
+        }
+
+        // Find any selected apps that meet the trigger
+        let triggerApp = null;
+        if (this._settings.get_strv(INHIBIT_APPS_KEY).length !== 0) {
+            switch (this._settings.get_enum(TRIGGER_APPS_MODE)) {
+            case AppsTrigger.ON_RUNNING:
+                triggerApp = this._findRunningApp();
+                break;
+            case AppsTrigger.ON_FOCUS:
+                triggerApp = this._findFocusedApp();
+                break;
+            case AppsTrigger.ON_ACTIVE_WORKSPACE:
+                triggerApp = this._findActiveApp();
+                break;
+            }
+        }
+
+        // Save any responsible app and update reasons
+        this._triggerApp = triggerApp;
+        if (triggerApp !== null) {
+            reasons.push('app');
+        }
+
+        return reasons;
+    }
+
+    _forceUpdate() {
+        // Remove any inhibitor, as settings / inhibit flags may have changed
+        if (this._isInhibited) {
+            this._removeInhibitor();
+        }
+
+        // Update state, as normal
+        this._updateState();
+    }
+
+    _updateState() {
+        // Get the reasons for inhibiting
+        let reasons = this._getInhibitReasons();
+
+        // If any ignored reasons are no longer active, stop ignoring them
+        this._ignoredReasons = this._ignoredReasons.filter(Set.prototype.has, new Set(reasons));
+
+        // Ignore any reasons in the ignore list, then save them
+        reasons = reasons.filter((n) => !this._ignoredReasons.includes(n));
+        this._lastReasons = [...reasons];
+        let shouldInhibit = reasons.length !== 0;
+
+        /* If no trigger app is running, and the light is in app-only mode,
+           and we left it blocked, unblock it
+        */
+        if (!reasons.includes('app')) {
+            if (this._settings.get_enum(NIGHT_LIGHT_KEY) === ControlContext.FOR_APPS) {
+                if (this._colorProxy.DisabledUntilTomorrow === true) {
+                    this._tempManageLight = true;
+                }
+            }
+        }
+
+        // Update inhibitor if required
+        if (this._isInhibited !== shouldInhibit) {
+            if (shouldInhibit) {
+                this._addInhibitor(reasons);
+            } else {
+                this._removeInhibitor();
+            }
+        }
+
```

Comment:
As it have only one inhibitor to manage, I recommend to destroy the previous inhibitor every time before creating the new one.  

I suggest to maybe store the cookie in a key (if it's possible) to make sure to remove it after the extension is badly stopped (crash). This can also be used to keep the inhibitor running after Caffeine been disabled/enabled when for example the screen is locked (it happened when using blank screen option).

Reply:
Mmm, I wondered about doing that, but I left it for now since it *shouldn't* be possible to incorrectly call them, and if they're being called at the wrong time we have a bug anyway.

I'll sort the current issues, then better protect the inhibitor for extra safety.